src/meta/src/controller/catalog.rs

```diff
@@ -0,0 +1,385 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::iter;
+
+use itertools::Itertools;
+use risingwave_common::catalog::{DEFAULT_SCHEMA_NAME, SYSTEM_SCHEMAS};
+use risingwave_pb::catalog::{PbDatabase, PbSchema};
+use risingwave_pb::meta::subscribe_response::{
+    Info as NotificationInfo, Operation as NotificationOperation,
+};
+use sea_orm::sea_query::{
+    Alias, CommonTableExpression, Expr, Query, QueryStatementBuilder, SelectStatement, UnionType,
+    WithClause,
+};
+use sea_orm::{
+    ActiveModelBehavior, ActiveModelTrait, ActiveValue, ColumnTrait, ConnectionTrait,
+    DatabaseConnection, DatabaseTransaction, EntityTrait, Iden, JoinType, ModelTrait, Order,
+    QueryFilter, Statement, TransactionTrait,
+};
+use tokio::sync::RwLock;
+
+use crate::controller::ObjectModel;
+use crate::manager::{DatabaseId, MetaSrvEnv, NotificationVersion, UserId};
+use crate::model_v2::object::ObjectType;
+use crate::model_v2::prelude::*;
+use crate::model_v2::{
+    connection, database, function, index, object, object_dependency, schema, sink, source, table,
+    view,
+};
+use crate::{MetaError, MetaResult};
+
+/// `CatalogController` is the controller for catalog related operations, including database, schema, table, view, etc.
+pub struct CatalogController {
+    env: MetaSrvEnv,
+    inner: RwLock<CatalogControllerInner>,
+}
+
+pub struct ReleaseContext {
+    streaming_jobs: Vec<i32>,
+    source_ids: Vec<i32>,
+    connections: Vec<serde_json::Value>,
+}
+
+impl CatalogController {
+    pub fn new(env: MetaSrvEnv) -> MetaResult<Self> {
+        let meta_store = env
+            .sql_meta_store()
+            .expect("sql meta store is not initialized");
+        Ok(Self {
+            env,
+            inner: RwLock::new(CatalogControllerInner {
+                db: meta_store.conn,
+            }),
+        })
+    }
+}
+
+struct CatalogControllerInner {
+    db: DatabaseConnection,
+}
+
+impl CatalogController {
+    async fn notify_frontend(
+        &self,
+        operation: NotificationOperation,
+        info: NotificationInfo,
+    ) -> NotificationVersion {
+        self.env
+            .notification_manager()
+            .notify_frontend(operation, info)
+            .await
+    }
+}
+
+impl CatalogController {
+    pub fn snapshot(&self) -> MetaResult<()> {
+        todo!("snapshot")
+    }
+
+    async fn create_object(
+        txn: &DatabaseTransaction,
+        obj_type: ObjectType,
+        owner_id: UserId,
+    ) -> MetaResult<object::Model> {
+        let mut active_db = object::ActiveModel::new();
+        active_db.obj_type = ActiveValue::Set(obj_type);
+        active_db.owner_id = ActiveValue::Set(owner_id as _);
```

Comment:
Can we use struct construction syntax for `ActiveModel`? This is to avoid missing fields if we are going to introduce new ones.

Reply:
The migration file itself is generated by sea-orm, but the detailed migration logic is hand-written. BTW, I defined the tables, indexes and foreign keys using dbdiagram and let chatgpt helped to write the codes. 🥵

Reply:
> Can we use struct construction syntax for ActiveModel? This is to avoid missing fields if we are going to introduce new ones.

👍  I think we'd better to construct it in this way, note that we still have to left some fields default and let the sql backend to fill them:
```rust
        let mut active_db = object::ActiveModel {
            oid: Default::default(),
            obj_type: ActiveValue::Set(obj_type),
            owner_id: ActiveValue::Set(owner_id as _),
            initialized_at: Default::default(),
            created_at: Default::default(),
        };
```